maskit/masks.py

```diff
@@ -388,6 +406,102 @@ def format_value(value):
+
+    @property
+    def mask(self) -> np.ndarray:
+        """
+        Accumulated mask of layers, wires, and parameters for both masking and freezing.
+        Note that this mask is readonly.
+        """
+        base = super().mask
+        base[self._parameter_freeze_mask.mask] = True
+        base[self._layer_freeze_mask, :] = True
+        base[:, self._wire_freeze_mask.mask] = True
+        return base
+
+    @property
+    def parameter_freeze_mask(self) -> Mask:
+        """Returns the encapsulated freezing parameter mask."""
+        return self._parameter_freeze_mask
+
+    @property
+    def wire_freeze_mask(self) -> Mask:
+        """Returns the encapsulated freezing wire mask."""
+        return self._wire_freeze_mask
+
+    @property
+    def layer_freeze_mask(self) -> Mask:
+        """Returns the encapsulated freezing layer mask."""
+        return self._layer_freeze_mask
+
+    def freeze(
+        self,
+        axis: PerturbationAxis = PerturbationAxis.LAYERS,
+        amount: Optional[Union[int, float]] = None,
+        mode: PerturbationMode = PerturbationMode.ADD,
+    ):
+        """
+        Freezes the parameter values for a given ``axis`` that is of type
+        :py:class:`~.PerturbationAxis`. The freezing is applied ``amount``times
+        and depends on the given ``mode`` of type :py:class:`~.PerturbationMode`.
+        If no amount is given, that is ``amount=None``, a random ``amount`` is
+        determined given by the actual size of the py:attr:`~.mask`. The ``amount``
+        is automatically limited to the actual size of the py:attr:`~.mask`.
+
+        :param amount: Number of items to freeze, defaults to None
+        :param axis: Which mask to freeze, defaults to PerturbationAxis.LAYERS
+        :param mode: How to freeze, defaults to PerturbationMode.ADD
+        :raises NotImplementedError: Raised in case of an unknown mode
+        """
+        assert mode in list(
+            PerturbationMode
+        ), f"The selected perturbation mode {mode} is not supported."
+        if amount == 0:
+            return
+        if axis == PerturbationAxis.LAYERS:
+            self._layer_freeze_mask.perturb(amount=amount, mode=mode)
+        elif axis == PerturbationAxis.WIRES:
+            self._wire_freeze_mask.perturb(amount=amount, mode=mode)
+        elif axis == PerturbationAxis.RANDOM:  # Axis is on parameters
+            self._parameter_freeze_mask.perturb(amount=amount, mode=mode)
+        else:
+            raise NotImplementedError(f"The perturbation {axis} is not supported")
+
+    def copy(self) -> "FreezableMaskedCircuit":
+        """Returns a copy of the current FreezableMaskedCircuit."""
+        clone = object.__new__(type(self))
+        clone._parameter_mask = self._parameter_mask.copy(clone)
+        clone._layer_mask = self._layer_mask.copy(clone)
+        clone._wire_mask = self._wire_mask.copy(clone)
+        clone.parameters = self.parameters.copy()
```

Comment:
You should be able to use ``clone = super().copy()`` for the first few lines here.

Reply:
Just realised that the annotations are not up to speed for my suggestion. 😓  We need to parameterise over "the type of ``self``" here.

We first need a type variable for that, ideally at the top of the module:
```python3
from typing import TypeVar
Self = TypeVar("Self")
```
Then, you can change *both* ``copy`` methods to say "whatever is the type of ``self`` is also the return type":
```python3
    def copy(self: Self) -> Self:
```